File: algorithms/algorithms-cplusplus/leetcode/Question_0226_Invert_Binary_Tree.cc
#include <queue>

using namespace std;

// Definition for a binary tree node.
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
 public:
  TreeNode* invertTreeRecursive(TreeNode* root) {
    if (!root) {
      return nullptr;
    }
    TreeNode* left = invertTree(root->left);
    TreeNode* right = invertTree(root->right);
    root->right = left;
    root->left = right;
    return root;
  }

  TreeNode* invertTree(TreeNode* root) {
    if (!root) {
      return nullptr;
    }
    queue<TreeNode*> q;
    q.push(root);
    while (!q.empty()) {
      TreeNode* cur = q.front();
      q.pop();
      TreeNode* temp = cur->right;
      cur->right = cur->left;
      cur->left = temp;
      if (cur->right) {
        q.push(cur->right);
      }
      if (cur->left) {
        q.push(cur->left);
      }
    }
    return root;
  }
};

File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0226_Invert_Binary_Tree.java
package com.lanceolata.coding.algorithms.java.leetcode;

import java.util.LinkedList;
import java.util.Queue;

public class Question_0226_Invert_Binary_Tree {
    // Definition for a binary tree node.
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode() {}
        TreeNode(int val) { this.val = val; }
        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    public TreeNode invertTreeRecursive(TreeNode root) {
        if (root == null) {
            return null;
        }
        TreeNode right = invertTreeRecursive(root.right);
        TreeNode left = invertTreeRecursive(root.left);
        root.left = right;
        root.right = left;
        return root;
    }

    public TreeNode invertTree2(TreeNode root) {
        if (root == null) {
            return null;
        }
        Queue<TreeNode> queue = new LinkedList<TreeNode>();
        queue.add(root);
        while (!queue.isEmpty()) {
            TreeNode current = queue.poll();
            TreeNode temp = current.left;
            current.left = current.right;
            current.right = temp;
            if (current.left != null) {
                queue.add(current.left);
            }
            if (current.right != null) {
                queue.add(current.right);
            }
        }
        return root;
    }
}
